lib/plugin.bash

```
@@ -26,37 +56,267 @@ function plugin_read_list() {
+  local status_code
+  status_code=$(echo "${invoke_output}" | jq -r '.StatusCode')
+
+  local response_payload
+  response_payload=$(cat "${response_file}")
+
+  if [[ "${status_code}" -ne "${expected_status}" ]]; then
+    log_error "Function invocation failed with status: ${status_code} (expected ${expected_status})"
+    log_info "Response payload:"
+    echo "${response_payload}"
+    rm -f "${response_file}"
+    return 1
+  fi
+
+  # Check for a business-logic error in the function's response.
+  # This occurs when the function itself traps an error and reports it, but still returns a 200 status code.
+  if echo "${response_payload}" | jq -e '.FunctionError' >/dev/null; then
+    local function_error
+    function_error=$(echo "${response_payload}" | jq -r '.FunctionError')
+    log_error "Function execution returned an error: ${function_error}"
+    log_info "Response payload:"
+    echo "${response_payload}"
+    rm -f "${response_file}"
+    return 1
+  fi
+
+  log_success "Function invocation successful (status: ${status_code})"
+  rm -f "${response_file}"
+  return 0
+}
+
+function create_lambda_function() {
+  local function_name="$1"
+  local alias_name="$2"
+  local aws_args=("${@:3}")
+
+  log_header ":construction: Creating Lambda function ${function_name}"
+
+  # Get required parameters
+  local package_type="${BUILDKITE_PLUGIN_AWS_LAMBDA_DEPLOY_PACKAGE_TYPE:-Zip}"
+  local runtime="${BUILDKITE_PLUGIN_AWS_LAMBDA_DEPLOY_RUNTIME:-python3.9}"
+  local handler="${BUILDKITE_PLUGIN_AWS_LAMBDA_DEPLOY_HANDLER:-lambda_function.lambda_handler}"
+  local role="${BUILDKITE_PLUGIN_AWS_LAMBDA_DEPLOY_ROLE?Missing role for function creation}"
+  local timeout="${BUILDKITE_PLUGIN_AWS_LAMBDA_DEPLOY_TIMEOUT:-30}"
+  local memory_size="${BUILDKITE_PLUGIN_AWS_LAMBDA_DEPLOY_MEMORY_SIZE:-128}"
+
+  local create_command=(aws lambda create-function "${aws_args[@]}" --function-name "${function_name}")
```

We might want to validate ourselves that the function name doesn't exceed the AWS length limit: https://docs.aws.amazon.com/lambda/latest/api/API_CreateFunction.html#API_CreateFunction_RequestBody

---

AWS would return an error anyway, but added it now, so it fails fast and points to the root cause of an issue.

---

EDIT: Looks like it's caught earlier anyway:
```
An error occurred (InvalidParameterValueException) when calling the CreateFunction operation: 1 validation error detected: Value 'test-canary-function-long-long-long-long-long-loooooooooooooooooooooooooooooooooooooooong' at 'functionName' failed to satisfy constraint: Member must have length less than or equal to 64
```

Will remove additional handling from the plugin's code then.